### PR TITLE
Immutable references

### PR DESCRIPTION
**Pull request by msladek**
https://synjira.atlassian.net/browse/CELDEV-1010

### CODE REVIEW

File: celements-reference/src/main/java/org/xwiki/model/internal/DefaultModelConfiguration.java [hunk 1]
```diff
@@ -72,10 +80,10 @@
   private ComponentManager componentManager;
 
   /**
-   * {@inheritDoc}
-   *
-   * @see org.xwiki.model.ModelConfiguration#getDefaultReferenceValue(org.xwiki.model.EntityType)
+   * The logger to log.
    */
+  private final Logger logger = LoggerFactory.getLogger(this.getClass());
```
Comment (fpichler) [inline celements-reference/src/main/java/org/xwiki/model/internal/DefaultModelConfiguration.java]: please make logger static -> naming: capital letters

**Review by fpichler — CHANGES_REQUESTED** [review]

File: celements-reference/src/main/java/org/xwiki/model/reference/DocumentReference.java [hunk 2]
```diff
@@ -22,90 +22,259 @@
...
 import org.xwiki.model.EntityType;
 
 /**
  * Represents a reference to a document (wiki, space and page names).
  *
- * @version $Id$
+ * @version $Id: d8ee9872095473617b4010f33b95ccc40da79b35 $
  * @since 2.2M1
  */
 public class DocumentReference extends EntityReference {
```
Comment (fpichler) [inline celements-reference/src/main/java/org/xwiki/model/reference/DocumentReference.java:38]: please add a static final serialVersionUID of type long (part of the Serializable contract)

File: celements-reference/src/main/java/org/xwiki/model/reference/EntityReference.java [hunk 3]
```diff
@@ -39,75 +49,268 @@
...
+   * @param name
+   *          the name of the parameter
+   * @param value
+   *          the value of the parameter
+   * @since 3.3M2
+   */
+  protected void setParameter(String name, Serializable value) {
+    if (value != null) {
+      if (parameters == null) {
+        parameters = new TreeMap<>();
```
Comment (fpichler) [inline celements-reference/src/main/java/org/xwiki/model/reference/EntityReference.java]: TreeMap is not thread safe. Thus calling setParameter outside of the constructor leads to not thread safe objects. I could not find any calls outside of custructors. Yet, maybe explain in JavaDoc or just change TreeMap to ConcurrentSkipListMap

**Review by fpichler — CHANGES_REQUESTED** [review]

File: celements-reference/src/main/java/org/xwiki/model/reference/EntityReference.java (hunk 3 above)
Comment (msladek) [reply celements-reference/src/main/java/org/xwiki/model/reference/EntityReference.java]: Strangely XWiki hasn't implemented immutability with final fields but with protected setter methods called in the constructor which are overwritten by subclasses. I don't know the reason for this design choice. I have refactored the `TreeMap` to a `final ImmutableMap` and removed the `setParameter` method.

File: celements-reference/src/main/java/org/xwiki/model/reference/DocumentReference.java [hunk 4]
```diff
@@ -23,17 +23,21 @@
...
 import java.util.List;
 import java.util.Locale;
+import java.util.Optional;
 
 import org.xwiki.model.EntityType;
 
+import one.util.streamex.EntryStream;
+
 /**
  * Represents a reference to a document (wiki, space and page names).
```
Comment (fpichler) [inline celements-reference/src/main/java/org/xwiki/model/reference/DocumentReference.java]: "(wiki, space and page names)" maybe "language" or Local should be added to this java comment, wdyt?

**Review by fpichler — CHANGES_REQUESTED** [review]

**Review by fpichler — APPROVED** [review]

**Pull request merged**